idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/editor/foldingbuilder/FoldingSyntax.kt

```diff
@@ -0,0 +1,46 @@
+package arrow.meta.ide.dsl.editor.foldingbuilder
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.lang.ASTNode
+import com.intellij.lang.Language
+import com.intellij.lang.folding.FoldingBuilder
+import com.intellij.lang.folding.FoldingDescriptor
+import com.intellij.lang.folding.LanguageFolding
+import com.intellij.openapi.editor.Document
+import com.intellij.util.KeyedLazyInstance
+import org.jetbrains.kotlin.idea.KotlinLanguage
+
+interface FoldingSyntax {
+  fun IdeMetaPlugin.languageFolding(
+    builder: FoldingBuilder,
+    lang: Language = KotlinLanguage.INSTANCE
+  ): ExtensionPhase =
+    extensionProvider(
+      LanguageFolding.EP_NAME,
+      object : KeyedLazyInstance<FoldingBuilder> {
+        override fun getKey(): String = lang.displayName
+
+        override fun getInstance(): FoldingBuilder = builder
+      }
+    )
+
+  fun IdeMetaPlugin.addFoldingBuilder(
+    placeHolderText: (node: ASTNode) -> String?,
+    foldRegions: (node: ASTNode, document: Document) -> Array<FoldingDescriptor>,
```

Comment:
```suggestion
    foldRegions: (node: ASTNode, document: Document) -> List<FoldingDescriptor>,
```
lists are simpler :)